README.rst

============
Python Money
============

Money class with optional CLDR-backed locale-aware formatting and an extensible currency exchange solution.

.. RADAR: version

**This is version 1.3.0**.

:Development: https://github.com/carlospalol/money
:Latest release: https://pypi.python.org/pypi/money/

This package is compatible with Python 2.7, 3.4, 3.5 (tested in 2.7.11, 3.4.4, 3.5.1), but there are important `Differences between Python versions`_. All code examples use Python 3.x.

**Contents**

.. contents::
    :local:
    :backlinks: none


Installation
============

Install the latest release with:

::

    pip install money

For locale-aware formatting, also install the latest version of `Babel <https://pypi.python.org/pypi/Babel>`_ (2.2 or higher required):

::

    pip install babel


Usage
=====

.. code:: python

    >>> from money import Money
    >>> m = Money(amount='2.22', currency='EUR')
    >>> m
    EUR 2.22

*amount* can be any valid value in ``decimal.Decimal(value)`` and *currency* should be a three-letter currency code. Money objects are immutable by convention and hashable. Once created, you can use read-only properties ``amount`` (decimal.Decimal) and ``currency`` (str) to access its internal components:

.. code:: python

    >>> m = Money(2, 'USD')
    >>> m.amount
    Decimal('2')
    >>> m.currency
    'USD'

Money emulates a numeric type and you can apply most arithmetic and comparison operators between money objects, as well as addition, subtraction, and division with integers (int) and decimal numbers (decimal.Decimal):

.. code:: python

    >>> m = Money('2.22', 'EUR')
    >>> m / 2
    EUR 1.11
    >>> m + Money('7.77', 'EUR')
    EUR 9.99

More formally, with *AAA* and *BBB* being different currencies:

+-----------+---------------+-----------+-----------+-----------------+
|           | Operator      | Money AAA | Money BBB | int, Decimal    |
+===========+===============+===========+===========+=================+
| **Money   | ``+``, ``-``  | Money     | N/A       | Money           |
+ AAA**     +---------------+-----------+-----------+-----------------+
|           | ``*``         | N/A       | N/A       | Money           |
+           +---------------+-----------+-----------+-----------------+
|           | ``/``, ``//`` | Decimal   | N/A       | Money           |
+           +---------------+-----------+-----------+-----------------+
|           | ``>``, ``>=`` | Compares  | N/A       | N/A             |
|           | ``<``, ``<=`` | amount.   |           |                 |
+           +---------------+           +-----------+-----------------+
|           | ``==``        |           | False     | False           |
|           |               |           |           |                 |
+-----------+---------------+-----------+-----------+-----------------+

Arithmetic operations with floats are not directly supported. If you need to operate with floats, you must first convert the float to a Decimal, or the Money object to a float (i.e. float(m)). Please be aware of the `issues and limitations of floating point arithmetics <https://docs.python.org/3/tutorial/floatingpoint.html>`_.


Currency presets
----------------

If you use fixed currencies in your code, you may find convenient to create currency-preset Money subclasses:

.. code:: python

    class EUR(Money):
        def __init__(self, amount='0'):
            super().__init__(amount=amount, currency='EUR')
    
    price = EUR('9.99')


Formatting
==========

Money objects are printed by default with en_US formatting and the currency code.

.. code:: python

    >>> m = Money('1234.567', 'EUR')
    >>> str(m)
    'EUR 1,234.57'

Use ``format(locale=LC_NUMERIC, pattern=None, currency_digits=True, format_type='standard')`` for locale-aware formatting with currency expansion. ``format()`` relies on ``babel.numbers.format_currency()``, and **requires Babel** 2.2 or higher to be installed.

.. code:: python

    >>> m = Money('1234.567', 'USD')
    >>> m.format('en_US')
    '$1,234.57'
    >>> m.format('es_ES')
    '1.234,57\xa0$'

The character ``\xa0`` is an unicode non-breaking space. If no locale is passed, Babel will use your system's locale. You can also provide a specific pattern to format():

.. code:: python

    >>> m = Money('-1234.567', 'USD')
    >>> # Regular US format:
    >>> m.format('en_US', '¤#,##0.00') 
    '-$1,234.57'
    >>> # Custom negative format:
    >>> m.format('en_US', '¤#,##0.00;<¤#,##0.00>')
    '<$1,234.57>'
    >>> # Spanish format, full currency name:
    >>> m.format('es_ES', '#,##0.00 ¤¤¤')
    '-1.234,57 dólares estadounidenses'
    >>> # Same as above, but rounding (overriding currency natural format):
    >>> m.format('es_ES', '#0 ¤¤¤', currency_digits=False)
    '-1235 dólares estadounidenses'

For more details on formatting see `Babel docs on currency formatting <http://babel.pocoo.org/en/latest/api/numbers.html#babel.numbers.format_currency>`_. To learn more about the formatting pattern syntax check out `Unicode TR35 <http://www.unicode.org/reports/tr35/tr35-numbers.html#Number_Format_Patterns>`_.

Currency exchange
=================

Currency exchange works by "installing" a **backend** class that implements the abstract base class (`abc <https://docs.python.org/3/library/abc.html>`_) ``money.exchange.BackendBase``. Its API is exposed through ``money.xrates``, along with setup functions ``xrates.install(pythonpath)``, ``xrates.uninstall()``, and ``xrates.backend_name``.

A simple proof-of-concept backend ``money.exchange.SimpleBackend`` is included:

.. code:: python

    from decimal import Decimal
    from money import Money, xrates

    xrates.install('money.exchange.SimpleBackend')
    xrates.base = 'USD'
    xrates.setrate('AAA', Decimal('2'))
    xrates.setrate('BBB', Decimal('8'))
    
    a = Money(1, 'AAA')
    b = Money(1, 'BBB')
    
    assert a.to('BBB') == Money('4', 'BBB')
    assert b.to('AAA') == Money('0.25', 'AAA')
    assert a + b.to('AAA') == Money('1.25', 'AAA')



XMoney
======

You can use ``money.XMoney`` (a subclass of Money), for automatic currency conversion while adding, subtracting, and dividing money objects (+, +=, -, -=, /, //). This is useful when aggregating lots of money objects with heterogeneous currencies. The currency of the leftmost object has priority.

.. code:: python

    from money import XMoney
    
    # Register backend and rates as above...
    
    a = XMoney(1, 'AAA')
    b = XMoney(1, 'BBB')
    
    assert sum([a, b]) == XMoney('1.25', 'AAA')


Exceptions
==========

Found in ``money.exceptions``.

``MoneyException(Exception)``
    Base class for all exceptions.

``CurrencyMismatch(MoneyException, ValueError)``
    Thrown when mixing different currencies, e.g. ``Money(2, 'EUR') + Money(2, 'USD')``. Money objects must be converted first to the same currency, or XMoney could be used for automatic conversion.

``InvalidOperandType(MoneyException, TypeError)``
    Thrown when attempting invalid operations, e.g. multiplication between money objects.

``ExchangeError(MoneyException)``
    Base class for exchange exceptions.

``ExchangeBackendNotInstalled(ExchangeError)``
    Thrown if a conversion is attempted, but there is no backend available.

``ExchangeRateNotFound(ExchangeError)``
    The installed backend failed to provide a suitable exchange rate between the origin and target currencies.


Hierarchy
---------

* ``MoneyException``
    * ``CurrencyMismatch``
    * ``InvalidOperandType``
    * ``ExchangeError``
        * ``ExchangeBackendNotInstalled``
        * ``ExchangeRateNotFound``



.. _python-differences:

Differences between Python versions
===================================

.. list-table::
    :header-rows: 1
    :stub-columns: 1
    
    * - Expression
      - Python 2.x
      - Python 3.x
    
    * - ``round(Money('2.5', 'EUR'))``
      - Returns ``3.0``, a **float** rounded amount **away from zero**.
      - Returns ``EUR 2``, a **Money object** with rounded amount to the **nearest even**.
    
    * - ``Money('0', 'EUR').amount < '0'``
      - Returns ``True``. This is the weird but expected behaviour in Python 2.x when comparing Decimal objects with non-numerical objects (Note the '0' is a string). `See note in docs <https://docs.python.org/2/library/stdtypes.html#comparisons>`_.
      - TypeError: unorderable types: decimal.Decimal() > str()



Design decisions
================

There are several design decisions in *money* that differ from currently available money class implementations:

Localization
------------

Do not keep any kind of locale conventions database inside this package. Locale conventions are extensive and change over time; keeping track of them is a project of its own. There is already such a project and database (the Unicode Common Locale Data Repository), and an excellent python API for it: `Babel <https://pypi.python.org/pypi/Babel>`_.

Currency
--------

There is no need for a currency class. A currency is fully identified by its ISO 4217 code, and localization or exchange rates data are expected to be centralized as databases/services because of their changing nature.

Also:

+ **Modulo operator (%)**: do not override to mean "percentage".
+ **Numeric type**: you **can** mix numbers and money in binary operations, and objects evaluate to False if their amount is zero.
+ **Global default currency**: subclassing is a safer solution.


Contributions
=============

Contributions are welcome. You can use the `regular github mechanisms <https://help.github.com/>`_.

To be forward-compatible, and given the small size of the package, Python 2.7 is supported in a different source "branch" at ``src-py2``.

To test your changes you will need `tox <https://pypi.python.org/pypi/tox>`_ and python 2.7, 3.4, and 3.5. Simply cd to the package root (by setup.py) and run ``tox``.


License
=======

money is released under the **MIT license**, which can be found in the file ``LICENSE``.






CHANGELOG

1.2.3
=====

2016-04-17

- Money 1.2.x supports Babel up to 2.1.x

1.2.2
=====

2014-06-15

- Support for SQLAlchemy composite columns (@dahlia)


1.2.1
=====

2014-06-15

- Fixed bug in __rsub__() (fix by zedlander)


1.2.0
=====

2014-04-30

- **Multiplication between a Money object and a float does not implicitly convert the float to a Decimal anymore.** (reported by spookylukey)


1.1.0 
=====

2014-04-01

- Added tox testing (requested by kvesteri)
- Compatible with Python 2.7, 3.3, 3.4 (requested by kvesteri)


1.0.2
=====

2013-11-25

- Initial stable version
- Compatible with Python 3.3




src/money/money.py
"""
Money classes
"""
import decimal
import re
from distutils.version import StrictVersion

from .exchange import xrates
from .exceptions import CurrencyMismatch, ExchangeRateNotFound


__all__ = ['Money', 'XMoney']

BABEL_AVAILABLE = False
BABEL_VERSION = None
REGEX_CURRENCY_CODE = re.compile("^[A-Z]{3}$")

try:
    import babel
    import babel.numbers
    BABEL_AVAILABLE = True
    BABEL_VERSION = StrictVersion(babel.__version__)
except ImportError:
    pass


class Money(object):
    """Money class with a decimal amount and a currency"""
    __hash__ = None
    
    def __init__(self, amount="0", currency=None):
        try:
            self.amount = decimal.Decimal(amount)
        except decimal.InvalidOperation:
            raise ValueError("amount value could not be converted to "
                             "Decimal(): '{}'".format(amount)) from None
        if currency in [None, False, '']:
            raise ValueError("invalid currency value: '{}'".format(currency))
        if not REGEX_CURRENCY_CODE.match(currency):
            raise ValueError("currency not in ISO 4217 format: "
                             "'{}'".format(currency))
        self.currency = currency
    
    def __repr__(self):
        return "{} {}".format(self.currency, self.amount)
    
    def __str__(self):
        return "{} {:,.2f}".format(self.currency, self.amount)
    
    def __lt__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '<')
            other = other.amount
        return self.amount < other
    
    def __le__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '<=')
            other = other.amount
        return self.amount <= other
    
    def __eq__(self, other):
        if isinstance(other, Money):
            return ((self.amount == other.amount) and
                    (self.currency == other.currency))
        return False
    
    def __ne__(self, other):
        return not self == other
    
    def __gt__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '>')
            other = other.amount
        return self.amount > other
    
    def __ge__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '>=')
            other = other.amount
        return self.amount >= other
    
    def __bool__(self):
        """
        Considering Money a numeric type (on ``amount``):
        
        bool(Money(2, 'XXX')) --> True
        bool(Money(0, 'XXX')) --> False
        """
        return bool(self.amount)
    
    def __add__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '+')
            other = other.amount
        amount = self.amount + other
        return self.__class__(amount, self.currency)
    
    def __radd__(self, other):
        return self.__add__(other)
    
    def __sub__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '-')
            other = other.amount
        amount = self.amount - other
        return self.__class__(amount, self.currency)
    
    def __rsub__(self, other):
        return (-self).__add__(other)
    
    def __mul__(self, other):
        if isinstance(other, Money):
            raise TypeError("multiplication is unsupported between "
                            "two money objects")
        amount = self.amount * other
        return self.__class__(amount, self.currency)
    
    def __rmul__(self, other):
        return self.__mul__(other)
    
    def __truediv__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '/')
            elif other.amount == 0:
                raise ZeroDivisionError()
            return self.amount / other.amount
        else:
            if other == 0:
                raise ZeroDivisionError()
            amount = self.amount / other
            return self.__class__(amount, self.currency)
    
    def __floordiv__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, '//')
            elif other.amount == 0:
                raise ZeroDivisionError()
            return self.amount // other.amount
        else:
            if other == 0:
                raise ZeroDivisionError()
            amount = self.amount // other
            return self.__class__(amount, self.currency)
    
    def __mod__(self, other):
        if isinstance(other, Money):
            raise TypeError("modulo is unsupported between two '{}' "
                            "objects".format(self.__class__.__name__))
        if other == 0:
            raise ZeroDivisionError()
        amount = self.amount % other
        return self.__class__(amount, self.currency)
    
    def __divmod__(self, other):
        if isinstance(other, Money):
            if other.currency != self.currency:
                raise CurrencyMismatch(self.currency, other.currency, 'divmod')
            elif other.amount == 0:
                raise ZeroDivisionError()
            return divmod(self.amount, other.amount)
        else:
            if other == 0:
                raise ZeroDivisionError()
            whole, remainder = divmod(self.amount, other)
            return (self.__class__(whole, self.currency),
                    self.__class__(remainder, self.currency))
    
    def __pow__(self, other):
        if isinstance(other, Money):
            raise TypeError("power operator is unsupported between two '{}' "
                            "objects".format(self.__class__.__name__))
        amount = self.amount ** other
        return self.__class__(amount, self.currency)
    
    def __neg__(self):
        return self.__class__(-self.amount, self.currency)
    
    def __pos__(self):
        return self.__class__(+self.amount, self.currency)
    
    def __abs__(self):
        return self.__class__(abs(self.amount), self.currency)
        
    def __int__(self):
        return int(self.amount)
    
    def __float__(self):
        return float(self.amount)
    
    def __round__(self, ndigits=0):
        return self.__class__(round(self.amount, ndigits), self.currency)
    
    def __composite_values__(self):
        return self.amount, self.currency
    
    def to(self, currency):
        """Return equivalent money object in another currency"""
        if currency == self.currency:
            return self
        rate = xrates.quotation(self.currency, currency)
        if rate is None:
            raise ExchangeRateNotFound(xrates.backend_name,
                                         self.currency, currency)
        amount = self.amount * rate
        return self.__class__(amount, currency)
    
    def format(self, locale=None, pattern=None):
        """
        Return a locale-aware, currency-formatted string.
        
        This method emulates babel.numbers.format_currency().
        
        A specific locale identifier (language[_territory]) can be passed,
        otherwise the system's default locale will be used. A custom
        formatting pattern of the form "¤#,##0.00;(¤#,##0.00)"
        (positive[;negative]) can also be passed, otherwise it will be
        determined from the locale and the CLDR (Unicode Common Locale Data
        Repository) included with Babel.
        
        >>> m = Money('1234.567', 'EUR')
        >>> m.format() # assuming the system's locale is 'en_US'
        €1,234.57
        >>> m.format('de_DE') # German formatting
        1.234,57 €
        >>> m.format('de', '#,##0 ¤') # German formatting (short), no cents
        1.235 €
        >>> m.format(pattern='#,##0.00 ¤¤¤') # Default locale, full name
        1,235.57 euro
        
        Learn more about this formatting syntaxis at:
        http://www.unicode.org/reports/tr35/tr35-numbers.html
        """
        if BABEL_AVAILABLE:
            if BABEL_VERSION >= StrictVersion('2.2'):
                raise Exception('Babel {} is unsupported. '
                    'Please upgrade Money to 1.3 or higher.'.format(
                    BABEL_VERSION))
            if not locale:
                locale = babel.default_locale('LC_NUMERIC')
            locale = babel.Locale.parse(locale)
            if not pattern:
                pattern = locale.currency_formats.get(pattern)
            pattern = babel.numbers.parse_pattern(pattern)
            return pattern.apply(self.amount, locale, currency=self.currency)
        else:
            raise NotImplementedError("formatting requires Babel "
                                      "(https://pypi.python.org/pypi/Babel)")
    
    @classmethod
    def loads(cls, s):
        """Parse from a string representation (repr)"""
        try:
            currency, amount = s.strip().split(' ')
            return cls(amount, currency)
        except ValueError as err:
            raise ValueError("failed to parse string '{}': "
                             "{}".format(s, err)) from None


class XMoney(Money):
    """Money subclass with implicit currency conversion"""
    def __lt__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__lt__(other)
    
    def __le__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__le__(other)
    
    def __gt__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__gt__(other)
    
    def __ge__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__ge__(other)
    
    def __add__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__add__(other)
    
    def __sub__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__sub__(other)
    
    def __truediv__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__truediv__(other)
    
    def __floordiv__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__floordiv__(other)
    
    def __divmod__(self, other):
        if isinstance(other, Money):
            other = other.to(self.currency)
        return super().__divmod__(other)







